### PR TITLE
feat(client): add filter params for model/dataset/runtime list command

### PR DESCRIPTION
**Pull request by dreamlandliu**
## Description
<a href="https://asciinema.org/a/KoiPQ7TzdYMpFuBX6CzdnnrMF" target="_blank"></a>

## Modules
- [ ] UI
- [ ] Controller
- [ ] Agent
- [x] Client
- [ ] Python-SDK
- [ ] Others

## Checklist
- [x] run code format and lint check
- [x] add unit test
- [ ] add necessary doc

### CODE REVIEW

**Comment by tianweidut** [comment]
@dreamlandliu add a demo show for pr description? for example: https://asciinema.org/a/545250

File: client/starwhale/base/bundle.py [hunk 1]
```diff
@@ -76,9 +76,10 @@ def list(
         project_uri: URI,
         page: int = DEFAULT_PAGE_IDX,
         size: int = DEFAULT_PAGE_SIZE,
+        _filter: t.Dict[str, t.Any] = {},
```
Comment (tianweidut) [inline client/starwhale/base/bundle.py:79]: Why use the private flag for the `_filter` argument?

File: client/starwhale/core/dataset/cli.py [hunk 2]
```diff
@@ -128,6 +128,9 @@ def _diff(
 @click.option(
     "--size", type=int, default=DEFAULT_PAGE_SIZE, help="Page size for dataset list"
 )
+@click.option("-n", "--name", help="Prefix of dataset name")
+@click.option("-o", "--owner", help="[Cloud]Name or id of the dataset owner")
+@click.option("-l", "--latest", is_flag=True, help="Only show the latest version")
```
Comment (tianweidut) [inline client/starwhale/core/dataset/cli.py:133]: Could we provide a filter option that can describe all filters syntax? for example: https://docs.docker.com/engine/reference/commandline/ps/#filtering

File: client/starwhale/core/dataset/model.py [hunk 3]
```diff
@@ -257,6 +257,7 @@ def list(
         project_uri: URI,
         page: int = DEFAULT_PAGE_IDX,
         size: int = DEFAULT_PAGE_SIZE,
+        _filter: t.Dict[str, t.Any] = {},
```
Comment (goldenxinxing) [inline client/starwhale/core/dataset/model.py:260]: How about add filter for local also?

File: client/starwhale/base/cloud.py [hunk 4]
```diff
@@ -235,23 +235,31 @@ def _fetch_bundle_all_list(
...
-            params={"pageNum": page, "pageSize": size},
+            params=_params,
             instance_uri=project_uri,
         ).json()
-
         objects = {}
 
+        _page = page
+        _size = size
+        if _filter.get("latest", False):
```
Comment (goldenxinxing) [inline client/starwhale/base/cloud.py:251]: [only for cloud] the api of 'list' are the collection for all runtime's latest info,  so the filter of 'latest' is no use in here in actual?

File: client/starwhale/base/bundle.py (hunk 1 above)
Comment (dreamlandliu) [reply client/starwhale/base/bundle.py]: done

File: client/starwhale/core/dataset/cli.py (hunk 2 above)
Comment (dreamlandliu) [reply client/starwhale/core/dataset/cli.py]: done

File: client/starwhale/base/bundle.py [hunk 5]
```diff
@@ -76,9 +76,29 @@ def list(
         project_uri: URI,
         page: int = DEFAULT_PAGE_IDX,
         size: int = DEFAULT_PAGE_SIZE,
+        filters: t.Union[t.Dict[str, t.Any], t.List[str]] = {},
```
Comment (tianweidut) [inline client/starwhale/base/bundle.py:79]: `filters: t.Union[t.Dict[str, t.Any], t.List[str]] = {},` 
-->
`filters: t.Optional[t.Union[t.Dict[str, t.Any], t.List[str]]] = None,` ?

File: client/starwhale/base/bundle.py [hunk 6]
```diff
@@ -76,9 +76,29 @@ def list(
...
     ) -> t.Tuple[t.Dict[str, t.Any], t.Dict[str, t.Any]]:
         _cls = cls._get_cls(project_uri)
-        return _cls.list(project_uri, page, size)  # type: ignore
+        _filter = cls.get_filter_dict(filters, cls.get_filter_fields())
+        return _cls.list(project_uri, page, size, _filter)  # type: ignore
+
+    @classmethod
+    def get_filter_dict(
+        cls, filters: t.Union[t.Dict[str, t.Any], t.List[str]], fields: t.List[str] = []
```
Comment (tianweidut) [inline client/starwhale/base/bundle.py:87]: fields that use mutable default value will meet the same issue as filters in 79 line.

Comment (tianweidut) [reply client/starwhale/base/bundle.py:87]: `fields: t.List[str] = []`
-->
`fields: t.Optional[t.List[str]] = None`
+ 
`fields = fields or None`

File: client/starwhale/base/bundle.py [hunk 7]
```diff
@@ -76,9 +76,29 @@ def list(
...
+    ) -> t.Dict[str, t.Any]:
+        if isinstance(filters, t.Dict):
+            return {k: v for k, v in filters.items() if k in fields}
+
+        _filter_dict = {}
+        for _f in filters:
+            _item = _f.split("=", 1)
+            if _item[0] in fields:
+                _filter_dict[_item[0]] = _item[1] if len(_item) > 1 else True
```
Comment (tianweidut) [inline client/starwhale/base/bundle.py:96]: Does `True` as the default value meet the unexpected scenario?

For example:
`--filter name` will be converted into `--filter name=True`, `True` is a bool value which is inappropriate for `name` value.

File: client/starwhale/base/cloud.py [hunk 8]
```diff
@@ -235,23 +235,31 @@ def _fetch_bundle_all_list(
         uri_typ: str,
         page: int = DEFAULT_PAGE_IDX,
         size: int = DEFAULT_PAGE_SIZE,
+        _filter: t.Dict[str, t.Any] = {},
```
Comment (tianweidut) [inline client/starwhale/base/cloud.py:238]: `_filter` -> `filters` ?

Comment (tianweidut) [reply client/starwhale/base/cloud.py:238]: In addition, the mutable default value as an argument is an anti-patten for python.

**Comment by tianweidut** [comment]
@dreamlandliu  updated docs? https://github.com/star-whale/starwhale/tree/main/docs/i18n/zh/docusaurus-plugin-content-docs/current/reference/cli

File: client/starwhale/base/bundle.py (hunk 5 above)
Comment (dreamlandliu) [reply client/starwhale/base/bundle.py]: done

File: client/starwhale/base/bundle.py (hunk 6 above)
Comment (dreamlandliu) [reply client/starwhale/base/bundle.py]: done

File: client/starwhale/base/bundle.py (hunk 7 above)
Comment (dreamlandliu) [reply client/starwhale/base/bundle.py]: done

File: client/starwhale/base/cloud.py (hunk 8 above)
Comment (dreamlandliu) [reply client/starwhale/base/cloud.py]: done

File: client/starwhale/core/dataset/model.py (hunk 3 above)
Comment (dreamlandliu) [reply client/starwhale/core/dataset/model.py]: done

**Review by tianweidut — APPROVED** [review]
LGTM

**Comment by tianweidut** [comment]
https://github.com/star-whale/starwhale/issues/1281

**Pull request merged**